src/pkcs11/framework-pkcs15.c

```diff
@@ -1446,8 +1487,17 @@ _add_public_objects(struct sc_pkcs11_slot *slot, struct pkcs15_fw_data *fw_data)
 		if (obj->base.flags & SC_PKCS11_OBJECT_SEEN)
 			continue;
 		/* Ignore 'private' object */
-		if (obj->p15_object->flags & SC_PKCS15_CO_FLAG_PRIVATE)
+		if (obj->p15_object->flags & SC_PKCS15_CO_FLAG_PRIVATE) {
+			/* If we found some non-accessible public object,
+			 * we can no longer claim Public Ceritificate Token conformance */
+			if ((obj->p15_object->type >= SC_PKCS15_TYPE_PUBKEY &&
+				obj->p15_object->type <= SC_PKCS15_TYPE_PUBKEY_EC) ||
+				(obj->p15_object->type >= SC_PKCS15_TYPE_CERT &&
+				obj->p15_object->type <= SC_PKCS15_TYPE_CERT_SPKI)) {
+				public_certificates = 0;
+			}
```

Comment:
I think it would be better to have an explicit check for all non-public objects here rather than relying on the specific order of OpenSC flags. You could, for example, use a switch statement with "fallthrough" cases

Reply:
Having better look, the `SC_PKCS15_TYPE_PUBKEY` and `SC_PKCS15_TYPE_CERT` are bitmasks of all the following types so the checking against the bitmask should be enough.